### ISSUE TITLE
Ability to optionally encode tuples

### ISSUE THREAD

**Issue by naps62** [michalmuskala/jason#89]
I'm often facing difficulties when using any kind of library that uses JSON encoding somewhere within it, because tuples may creep in within the argumnets

The main example that comes to mind, and that causes me the most trouble is https://github.com/getsentry/sentry-elixir

Patterns like this are very common:

```
case some_stuff() do
  {:ok, result} ->
    some_more_stuff()

 {:error, error} ->
   Sentry.capture_message("bla bla bla", extra: error)
end
```

Now, depending on what kind of error comes in, a tuple may be somewhere in there. Maybe this is a Phoenix conn object, or a failed Tesla http request, or some random object developed by someone else on my team. All of those can contain tuples within them, resulting in the JSON-encoding step failing, and me getting a error about that exception on Sentry's dashboard, instead of the actual error that was suposed to be tracked.

For this reason, and probably specifically to debugging-related operations such as error reporting, it could be useful to specify a custom way to encode tuples.

The only way to do that right now would be to do `defimpl Jason.Encoder, for: Tuple`. But this would be a global change, which is probably not a good idea, as it would hide actual errors (I don't want ending up giving back invalid results through my API responses)

It seems there's already a `:maps` option that can be sent along `Jason.encode/1` to switch between strict and pretty mode.

My proposal would be to add a similar `:tuples` option. Let's say for the sake of argument that `tuples: :debug` would do the following:

```
["tuple" | Tuple.to_list(tuple)]
```

This would allow me to create the following module:

```
defmodule MyCustomJason do
  def encode(value), do:
    Jason.encode(value, tuples: :debug)
  end
end
```

And then specify `MyCustomJason` as the json encoder to be used by Sentry

Would this be something possible to include in this library?

Note: I'm very much willing to do the PR myself, but I'd prefer to validate the idea first.

**Comment by michalmuskala** [comment]
I'm pretty sure as soon as you'd start encoding tuples there, you'll find many, many more things that couldn't be properly encoded - pids, refs, ports, funs, other structs, etc. I think there has to be some additional pass over the data anyway to somehow encode those before it's sent over the wire to report errors. Doing this for tuples would only be the first, somewhat limited, layer.

**Comment by naps62** [comment]
hmm, I see your point. I guess I was only thinking of tuples because it's by far the case that has happened the most in my latest projects.

Would any kind of more generic approach be possible? (I'm guessing not, but oh well, doesnt' hurt to ask).

Our current approach is to use the following:

```
  defp prevent_invalid_json(opts) do
    Keyword.update(opts, :extra, %{}, fn extra ->
      case Jason.encode(extra) do
        {:ok, _} -> extra
        _ -> %{invalid_json: true, original: inspect(extra)}
      end
    end)
  end
```

because this happened so many times, I found it easier to just fallback to `inspect` any time there's an encoding error, because at least that way we get the error report we intended, even though the extra data is not as useful (if it contains sensitive data, then Sentry will filter out the whole string instead of just the bad keys.

This is only half-useful though. It helps me get more visibility into the error regardless of the encoding error, but I then need to first patch this particular case so that next time the error is properly reported
